crates/telio-wg/src/pq.rs

```diff
@@ -0,0 +1,298 @@
+const CIPHERTEXT_LEN: u32 = kyber768::ciphertext_bytes() as _;
+
+/// The PQ module error type
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    /// IO error
+    #[error("IO: {0:?}")]
+    Io(#[from] io::Error),
+    /// Generic unrecoverable error
+    #[error("Generic: {0}")]
+    Generic(String),
+}
+
+impl From<String> for Error {
+    fn from(value: String) -> Self {
+        Self::Generic(value)
+    }
+}
+
+impl From<&str> for Error {
+    fn from(value: &str) -> Self {
+        Self::Generic(value.to_string())
+    }
+}
+
+impl From<tokio::time::error::Elapsed> for Error {
+    fn from(value: tokio::time::error::Elapsed) -> Self {
+        Self::Io(io::Error::new(io::ErrorKind::TimedOut, value))
+    }
+}
+
+/// Post quantum keys retrived from hanshake
+pub struct PqKeys {
+    /// Kyber shared secret
+    pq_shared: kyber768::SharedSecret,
+    /// X25519 secret key
+    wg_secret: x25519_dalek::StaticSecret,
+}
+
+impl PqKeys {
+    /// Retrieve the PQ preshared key
+    pub fn pq_shared(&self) -> telio_crypto::PresharedKey {
+        #[allow(clippy::expect_used)]
+        telio_crypto::PresharedKey(
+            self.pq_shared
+                .as_bytes()
+                .try_into()
+                .expect("Preshared WG keys should have the same lenght as PQ preshared key"),
+        )
+    }
+
+    /// Retrieve the WG private key
+    pub fn wg_secret(&self) -> telio_crypto::SecretKey {
+        telio_crypto::SecretKey::new(self.wg_secret.to_bytes())
+    }
+}
+
+/// Get PQ keys from the VPN server
+pub async fn fetch_keys(
+    endpoint: impl ToSocketAddrs,
+    secret: &telio_crypto::SecretKey,
+    peers_pubkey: &telio_crypto::PublicKey,
+) -> Result<PqKeys, Error> {
+    let (tunn, sock) = handshake(endpoint, secret, peers_pubkey).await?;
+
+    let mut rng = rand::rngs::StdRng::from_entropy();
+
+    // Generate keys
+    let wg_secret = x25519_dalek::StaticSecret::random_from_rng(&mut rng);
+    let (pq_public, pq_secret) = kyber768::keypair();
+    let wg_public = x25519_dalek::PublicKey::from(&wg_secret);
+
+    const IPV4_HEADER_LEN: usize = 20;
+    const UDP_HEADER_LEN: usize = 8;
+    let mut pkgbuf = Vec::with_capacity(1024 * 4); // 4 KiB
+
+    pkgbuf.resize(IPV4_HEADER_LEN + UDP_HEADER_LEN, 0);
+
+    // Pepare payload
+    let method = 0u32; // get
+
+    // UDP packet payload
+    pkgbuf.extend_from_slice(&HANDSHAKE_VERSION.to_le_bytes());
+    pkgbuf.extend_from_slice(&method.to_le_bytes());
+    pkgbuf.extend_from_slice(&(wg_public.as_bytes().len() as u32).to_le_bytes());
+    pkgbuf.extend_from_slice(&wg_public.to_bytes());
+    pkgbuf.extend_from_slice(&(pq_public.as_bytes().len() as u32).to_le_bytes());
+    pkgbuf.extend_from_slice(pq_public.as_bytes());
+
+    let pkg_len = pkgbuf.len();
+
+    #[allow(clippy::expect_used)]
+    #[allow(index_access_check)]
+    let mut udppkg = MutableUdpPacket::new(&mut pkgbuf[IPV4_HEADER_LEN..])
+        .expect("UDP buffer should not be too small");
```

Comment:
We should eliminate unwraps and expects. It probably means that something went very bad however we can handle it easily by wrappig it in an error and allowing graceful retry with non-pq for example

Reply:
This is statically guaranteed to never panic

Reply:
Could you elaborate a bit more on this?

Reply:
According to the `MutableUdpPacket::new` docs, it panics only when the buffer provided is smaller than the UDP header size. The buffer used has at least the required length (the `UDP_HEADER`) variable